UI: Expose hooks that retrieve the highest alerts for each entity

ChengYanJin opened the pull request:
**Component**: ui, alerts

**Context**: 
Provide hooks to fetch the highest severity alerts for each entity to display the health.
```
 useNodesAlert(): Alert?
 useVolumesAlert(): Alert?
 useServiceAlert(): Alert?
 useNetworkAlert(): Alert?
 getNodesCountQuery(): Query // type Query = {cacheKey, queryFn, options} from react-query
 getVolumesCountQuery(): Query // type Query = {cacheKey, queryFn, options} from react-query
```
**Summary**:
```js
const nodesAlerts = useHighestSeverityAlerts(getNodesAlertName());
const volumesAlerts = useHighestSeverityAlerts(getVolumesAlertName());
const volumesAlerts = useHighestSeverityAlerts(getNetworksAlertName());
const volumesAlerts = useHighestSeverityAlerts(getServicesAlertName());
const coutNodes = getNodesCountQuery();
const coutVolumes = getVolumesCountQuery();
```

As discussed with @JBWatenbergScality, we will eventually remove the AlertProvider from Metalk8s UI.

**Acceptance criteria**: 


---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #3257

<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

bert-e commented:
# Hello chengyanjin,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nicolast

* @chengyanjin

* @alexandre-allard-scality

* @thomasdanan

* @gdemonet

* @teddyandrieux

* @lucieleonard

* @jbwatenbergscality

JBWatenbergScality commented on ui/src/services/alertHooks.js:
```diff
@@ -0,0 +1,69 @@
+/*
+The critical and warning alerts may active at the same time.
+how to use:
+  const nodesAlert = useHighestSeverityAlert({
```
Nit : You may want to use jsdoc notation for this kind of comments to benefit from it in IDE autocompletion 

The name of the functions doesn't match the implementation :
```suggestion
  const nodesAlert = useHighestSeverityAlerts({
```

JBWatenbergScality commented on ui/src/containers/AlertProvider.js:
```diff
@@ -22,7 +22,8 @@ export function useAlerts(filters: FilterLabels) {
 
-const AlertProvider = ({ children }: any) => {
+// the `fakequery` is for testing purpose in order to inject alerts
+const AlertProvider = ({ children, fakequery }: any) => {
```
```suggestion
import type {Node} from "react";
const AlertProvider = ({ children }: {children: Node}) => {
```

Instead of introducing a `fakeQuery` here I would go for adding an optional second parameter to useAlerts, this optional second parameter would be the alert context. Doing so in the test you won't need to render an AlertProvider wrapper anymore and could `createContext` initialised with your mocked query.

JBWatenbergScality replied:
Another alternative (maybe better because in this way it won't impact the definition of useAlerts neither useHighestSeverityAlerts definition) could be to move the context to a separate module, mock with jest.mock and returning a context containing your fakeQuery

ChengYanJin replied:
Sure. thanks for the suggestion. Will go for the second solution!

JBWatenbergScality commented on ui/src/services/alertHooks.js:
```diff
@@ -0,0 +1,69 @@
+
+type Query = { cacheKey: string, queryFn: Function, options?: Object };
+
+export const getNodesAlertName = (): string[] => {
+  return ['NodeAtRisk', 'NodeDegraded'];
+};
+
+export const getVolumesAlertName = (): string[] => {
+  return ['VolumeAtRisk', 'VolumeDegraded'];
+};
+
+export const getNetworksAlertName = (): string[] => {
+  return ['ControlPlaneNetworkDegraded', 'WorkloadPlaneNetworkDegraded'];
+};
+
+export const getServicesAlertName = (): string[] => {
+  return ['PlatformServicesAtRisk', 'PlatformServicesDegraded'];
+};
```
I think a better abstraction would be to return a set of selectors instead of the name, in this way if in the futur we need to narrow down the selection of alerts we won't need to change the definition of those functions.

 ```suggestion
export const getNodesAlertSelectors = (): FilterLabels => {
  return {name: ['NodeAtRisk', 'NodeDegraded']};
};
// ...
```

JBWatenbergScality commented on ui/src/services/alertHooks.js:
```diff
@@ -0,0 +1,69 @@
+import type { FilterLabels, Alert } from './alertUtils';
+import { getHealthStatus } from './alertUtils';
+
+type Query = { cacheKey: string, queryFn: Function, options?: Object };
```
It would be better to reuse Query type from react query

ChengYanJin committed:
shell-ui: Add alerthook and expose in library

Refs: #3257

ChengYanJin committed:
shell-ui: Add k8s service

Check if the k8s call carry the `id_token`, otherwise listen to the
navbarElement AUTHENTICATED_EVENT

Refs: #3275

ChengYanJin committed:
shell-ui: Add test for getNodesCount

Refs: #3257

JBWatenbergScality commented on shell-ui/src/platform/service/k8s.js:
```diff
@@ -0,0 +1,113 @@
+        .then((res) => {
+          return res.items.length;
+        }),
+    options: { refetchInterval: 10000 },
```
I think it is missing enabled options :

```suggestion
    options: { refetchInterval: 10000, enabled: token },
```

this will only triggers the query if the token is provided.

ChengYanJin replied:
Right, I've checked the doc, the `enabled` is a boolean, so maybe the following one is more correct
 ```options: { refetchInterval: 10000, enabled: token ? true : false },```

JBWatenbergScality replied:
JS is automatically casting this to a boolean so  `token ? true : false` is strictly equivalent to `token`. Actually in `token ? true : false` you are also casting token as a boolean for the ternary condition

ChengYanJin replied:
For some reason, the tests won't execute by `enabled: token`. So I have to switch back to `enabled: token ? true : false`

JBWatenbergScality commented on shell-ui/src/platform/service/k8s.js:
```diff
@@ -0,0 +1,113 @@
+  k8sUrl: string,
+  token?: string | null,
+): typeof useQuery => {
+  if (!token) {
+    console.error('K8s API Not authenticated');
+    return;
+  }
```
IMO it would be better to get rid of this early return in order to have a consistant return type even when the token is not provided. Ensuring the token is provided to perform the request can be done through the `enabled` useQuery option

JBWatenbergScality commented on shell-ui/src/platform/service/k8s.js:
```diff
@@ -0,0 +1,113 @@
+  k8sUrl: string,
+  token?: string | null,
+): typeof useQuery => {
+  if (!token) {
+    console.error('K8s API Not authenticated');
+    return;
+  }
```
same as above

JBWatenbergScality commented on shell-ui/src/platform/service/k8s.js:
```diff
@@ -0,0 +1,113 @@
+        .then((res) => {
+          return res.items.length;
+        }),
+    options: { refetchInterval: 10000 },
```
same as above

JBWatenbergScality commented on shell-ui/src/platform/service/k8s.js:
```diff
@@ -0,0 +1,113 @@
+  }, []);
+
+  const queryVolumesResult = useQuery(getVolumesCountQuery(k8sUrl, token));
+  queryVolumesResult.volumesCount = queryVolumesResult.data;
```
in the current implementation `queryVolumesResult`  can be null, see bellow my comment about the `enabled` option.

JBWatenbergScality commented on shell-ui/src/platform/service/k8s.js:
```diff
@@ -0,0 +1,113 @@
+  }, []);
+
+  const queryNodesResult = useQuery(getNodesCountQuery(k8sUrl, token));
+  queryNodesResult.nodesCount = queryNodesResult.data;
```
in the current implementation `queryNodesResult `  can be null, see bellow my comment about the `enabled` option.

JBWatenbergScality reviewed (commented):
-

JBWatenbergScality commented on shell-ui/src/alerts/services/alertHooks.js:
```diff
@@ -0,0 +1,40 @@
+ * @returns An array of alerts with the highest severity
+ */
+export const useHighestSeverityAlerts = (filters: FilterLabels): Alert[] => {
+  const query = useAlerts(useContext)(filters);
```
useContext should be injected

JBWatenbergScality commented on shell-ui/src/alerts/services/alertHooks.test.js:
```diff
@@ -0,0 +1,112 @@
+  afterEach(() => server.resetHandlers());
+
+  const alertLibrary = window.shellUIAlerts[version];
+  console.log('window.shellUIAlerts', window.shellUIAlerts);
```
It could be removed

ChengYanJin committed:
shell-ui: Add enabled option in query object

To disable the query if token is not provided

Refs: #3257

JBWatenbergScality commented:
> Cannot find module 'axios' from 'src/platform/service/k8s.js'

Axios is not available in shell-ui. I think we can simply use fetch as it is available in the browser we target

JBWatenbergScality commented on shell-ui/src/platform/service/k8s.js:
```diff
@@ -0,0 +1,113 @@
+  return {
+    cacheKey: 'countNodes',
+    queryFn: () =>
+      fetch(`${k8sUrl}/api/v1/nodes`)
```
missing authorization header :
```
{
        headers: {
          authorization: `bearer ${token}`,
        },
      }
```

JBWatenbergScality commented on shell-ui/src/platform/service/k8s.js:
```diff
@@ -0,0 +1,113 @@
+    queryFn: () =>
+      fetch(`${k8sUrl}/api/v1/persistentvolumes`, {
+        headers: {
+          authorisation: `bearer ${token}`,
```
it should be `authorization`

JBWatenbergScality commented:
I think you can merge this manually in my branch `improvement/shell-ui-alert-provider`

JBWatenbergScality commented on shell-ui/src/platform/service/k8s.js:
```diff
@@ -0,0 +1,110 @@
+        .then((res) => {
+          return res.items.length;
+        }),
+    options: { refetchInterval: 10000, enabled: token },
```
not sure about this options to be checked
